drivers/dma/dma_intel_adsp_hda.c

```diff
@@ -349,6 +353,42 @@ int intel_adsp_hda_dma_stop(const struct device *dev, uint32_t channel)
 		return 0;
 	}
 
+	/*
+	 * Workaround for random metallic noise on DMA restart:
+	 * Make sure that the output link DMA read and write pointers are exactly 0 when we disable
+	 * the channel.
+	 * On channel restart the pointers are not rested to 0 and can result corruption.
+	 */
+	if (cfg->direction == MEMORY_TO_PERIPHERAL) {
+		uint32_t wp, size;
+		int i;
+
+		wp = *DGBWP(cfg->base, cfg->regblock_size, channel);
+		size = intel_adsp_hda_get_buffer_size(cfg->base, cfg->regblock_size, channel);
+
+		/* Move the WP to wrap to exactly to 0 if it is not already 0 */
+		if (wp) {
+			intel_adsp_hda_link_commit(cfg->base, cfg->regblock_size, channel,
+						   size - wp);
+		}
+
+		/* Wait for the pointers to reach 0 */
+		for (i = 0; i < 1000; i++) {
+			uint32_t unused = intel_adsp_hda_unused(cfg->base, cfg->regblock_size,
+								channel);
+
+			if (unused >= size)
+				break;
+			k_sleep(K_USEC(1));
+		}
+
+		if (i >= 1000) {
```

Comment:
Replacing the above loop with a WAIT_FOR would let you do...

```c
if(!WAIT_FOR(intel_adsp_hda_unused(...) >= size, 1000, k_yield())) {
  LOG_WRN("Drain timed out...", ...);
}
```

WAIT_FOR implementation is here https://github.com/zephyrproject-rtos/zephyr/blob/9c0f92db475f0e27673ac366276026eddf1a150c/include/zephyr/sys/util.h#L837

Reply:
@teburd, the strange thing is that when I used WAIT_FOR then we had quite frequent timeout prints, I think @kv2019i might have looked at it and might have explanation.